File: translator/demo/clojure/lam1.clj
;; λ-1 translator — Clojure prelude
(def _failures (atom 0))

(defn encodeInt [n]              ; host int -> チャーチ数
  (fn [f] (fn [x] (reduce (fn [acc _] (f acc)) x (range n)))))

(defn decodeInt [t]             ; チャーチ数 -> 文字列
  (str ((t inc) 0)))

(defn decodeBool [t]            ; チャーチ真偽値 -> "true"/"false"
  ((t "true") "false"))

(defn _check [a b label]
  (if (= a b)
    (println (str "ok   " label))
    (do (swap! _failures inc)
        (println (str "FAIL " label ": " (pr-str a) " != " (pr-str b))))))

(defn _finish []
  (when (> @_failures 0)
    (println (str @_failures " failure(s)"))
    (System/exit 1))
  (println "all green"))

File: translator/demo/ruby/lam1.rb
# λ-1 translator — Ruby prelude
$failures = 0

def encodeInt(n)               # host int -> チャーチ数
  ->(f) { ->(x) { acc = x; n.times { acc = f.(acc) }; acc } }
end

def decodeInt(t)               # チャーチ数 -> 文字列
  t.(->(k) { k + 1 }).(0).to_s
end

def decodeBool(t)              # チャーチ真偽値 -> "true"/"false"
  t.('true').('false')
end

def _check(a, b, label)
  if a == b
    puts "ok   #{label}"
  else
    $failures += 1
    puts "FAIL #{label}: #{a.inspect} != #{b.inspect}"
  end
end

def _finish
  if $failures > 0
    puts "#{$failures} failure(s)"
    exit 1
  end
  puts "all green"
end
